### PR TITLE
fix: Internal Sharding Disconnects

### PR DESCRIPTION
**Pull request by Lewdcario**
**Please describe the changes this PR makes and why it should be merged:**
Main point is to get a shot at refactoring the problematic code that arose from constantly patching over the IS reconnect logic. Here, I try to take a different approach to spawning and resuming shards. The main idea was to fix the silent disconnects that have plagued a few bots running on recent master branch commits.
Did a little bit of a consistency's touch as far as event names go, as well.
This has been tested extensively, for a very long time span, and it seems to be working now. Any feedback/suggestions are greatly appreciated.

**Status**
- [x] Code changes have been tested against the Discord API, or there are no code changes
- [x] I know how to update typings and have done so, or typings don't need updating

**Semantic versioning classification:**  
- [x] This PR changes the library's interface (methods or parameters added)
  - [x] This PR includes breaking changes (methods removed or renamed, parameters moved or removed)
- [ ] This PR **only** includes non-code changes, like changes to documentation, README, etc.

### CODE REVIEW

**Comment by Lewdcario** [comment]
> How does this relate to #2976? It looks like they cover a lot of the same ground.

Most of the cleaning up is similar, but we took a different route in how the reconnects and invalid sessions were handled.

**Comment by CrazyPython** [comment]
Does #2992 reproduce with this PR?

File: src/client/websocket/WebSocketShard.js [hunk 1]
```diff
@@ -227,28 +249,23 @@ class WebSocketShard extends EventEmitter {
...
-
     if (packet.s > this.sequence) this.sequence = packet.s;
 
     switch (packet.op) {
       case OPCodes.HELLO:
         this.identify();
         return this.heartbeat(packet.d.heartbeat_interval);
       case OPCodes.RECONNECT:
-        return this.reconnect();
+        // Triggers a websocket recreation since this.expectingCLose won't be filled
```
Comment (amishshah) [inline src/client/websocket/WebSocketShard.js]: Should be `this.expectingClose`

**Comment by amishshah** [comment]
See https://github.com/discordjs/discord.js/commit/793341dbb401f2956d60fd6b9f94a611cf539126

**Comment by DougTy** [comment]
Unfortunately, 793341d does not fix my disconnects, whereas this did. I'll try and get more info but I'm not sure if I'll be able to find anything useful. After updating to the latest master and no longer using this PR, the disconnects are back. Is it possible that this is doing something that #2976 is not?

**Comment by vladfrangu** [comment]
That... Shouldn't happen at all. Can you add a `client.on('debug', console.log` or similar and see what gets returned?

**Comment by DougTy** [comment]
Will do, but it may take some time as this only happens once every day or two. The bot simply disconnects and never reconnects, just gets stuck at the reconnecting event.

**Comment by DougTy** [comment]
@vladfrangu This is what I get from debug:
```
[Shard 0] Sending a heartbeat
[Shard 0] Heartbeat acknowledged, latency of 61ms
[Shard 0] WebSocket was closed.
      Event Code: 1006
      Reason: 
[Shard 0] Clearing heartbeat interval
[2019-02-12 05:23:23]  event: reconnecting
[Shard 0] Reconnecting in 3500ms to the gateway...
[Shard 0] Connecting to wss://gateway.discord.gg/
```
It never progresses beyond that. Has been stuck there for a few hours now.

**Pull request closed without merging**